File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/MeleeWeaponTypeEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum MeleeWeaponTypeEnum implements AttackProficiencyEnum {
    SWORD("sword"),
    SPEAR("spear"),
    HAMMER("hammer"),
    POLEARM("polearm"),
    KNIFE("knife"),
    CLUB("club"),
    FLAIL("flail"),
    BRAWLING("brawling"),
    AXE("axe"),
    PICK("pick"),
    SHIELD("shield"),
    WHIP("whip");

    private final String value;
    private final static Map<String, MeleeWeaponTypeEnum> CONSTANTS = new HashMap<>();

    static {
        for (MeleeWeaponTypeEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    MeleeWeaponTypeEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static MeleeWeaponTypeEnum fromValue(String value) {
        MeleeWeaponTypeEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/MeleeWeaponTypeEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class MeleeWeaponTypeEnum(private val value: String) : AttackProficiencyEnum {
    SWORD("sword"), SPEAR("spear"), HAMMER("hammer"), POLEARM("polearm"), KNIFE("knife"), CLUB("club"), FLAIL("flail"), BRAWLING("brawling"), AXE("axe"), PICK("pick"), SHIELD("shield"), WHIP("whip");

    companion object {
        private val CONSTANTS: MutableMap<String, MeleeWeaponTypeEnum> = HashMap()
        @JsonCreator
        fun fromValue(value: String?): MeleeWeaponTypeEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[ai.ragu.charactermanager.enumeration.c.value] = ai.ragu.charactermanager.enumeration.c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}